tagging and gauge operations

gandarez opened the pull request:
* add support for tagging
* add support for gauge operations

Related to the issues [175](https://github.com/justeat/JustEat.StatsD/issues/175) and [176](https://github.com/justeat/JustEat.StatsD/issues/176).

martincostello commented on src/JustEat.StatsD/Buffered/StatsDMessage.cs:
```diff
@@ -1,31 +1,39 @@
 
-        private StatsDMessage(string statBucket, double magnitude, StatsDMessageKind messageKind)
+        private StatsDMessage(string statBucket, double magnitude, StatsDMessageKind messageKind,
+            IDictionary<string, string>? tags, Operation operation = default)
```
Nit: one parameter per line if not all on the same line please.

```suggestion
        private StatsDMessage(
            string statBucket,
            double magnitude,
            StatsDMessageKind messageKind,
            IDictionary<string, string>? tags,
            Operation operation = default)
```

martincostello commented on src/JustEat.StatsD/Buffered/BufferBasedStatsDPublisher.cs:
```diff
@@ -31,19 +32,21 @@ internal BufferBasedStatsDPublisher(StatsDConfiguration configuration, IStatsDTr
-        public void Gauge(double value, string bucket)
+        public void Gauge(double value, string bucket,
+            Operation operation = Operation.Set,
+            IDictionary<string, string>? tags = null)
```
Nit: one parameter per line if not on the same line.

```suggestion
        public void Gauge(
            double value,
            string bucket,
            Operation operation = Operation.Set,
            IDictionary<string, string>? tags = null)
```

martincostello commented:
Thank you for you contribution @gandarez!

Tags are fairly self-explanatory, but can you explain a little more about what the gauge changes are for please?

gandarez commented on tests/JustEat.StatsD.Tests/IntegrationTests.cs:
```diff
@@ -53,16 +53,16 @@ private static async Task AssertMetrics(StatsDConfiguration config, IStatsDPubli
 
             // Act - Create a gauge
-            publisher.Gauge(3.141, "circle");
+            publisher.Gauge(3.141, "circle", Operation.Set, null);
```
Is there an easy way or at least documented on how to run integration tests locally?

gandarez commented on src/JustEat.StatsD/PublicAPI.Shipped.txt:
```diff
@@ -14,9 +14,13 @@ JustEat.StatsD.IDisposableTimer
 JustEat.StatsD.IDisposableTimer.Bucket.get -> string
 JustEat.StatsD.IDisposableTimer.Bucket.set -> void
 JustEat.StatsD.IStatsDPublisher
-JustEat.StatsD.IStatsDPublisher.Gauge(double value, string bucket) -> void
```
Should I move those lines to `Unshipped`?

martincostello replied:
They're fine here rather than in the other file, but the need to change the signatures at all highlights the binary breaking nature of these changes.

gandarez commented:
@martincostello thanks!

Fo all the comments you've done I already fixed them locally. There are only two left I'd need your assistance. The first is about the `Shipped.txt` and the other is how to I start integration tests in my machine.

I followed the documentation for gauges and implemented the concept of being able to change the sign for a gauge.
<https://github.com/statsd/statsd/blob/master/docs/metric_types.md#gauges>

gandarez commented:
@martincostello please be adivised that I needed to add `keyNameSanitize: false` to avoid the server to remove the `semicolon` and `equal` signs when tagging.

martincostello commented on src/JustEat.StatsD/Buffered/StatsDUtf8Formatter.cs:
```diff
@@ -79,7 +99,15 @@ private static bool TryWritePayloadWithMessageKindSuffix(ref Buffer buffer, in S
                         bool isMagnitudeIntegral = msg.Magnitude == integralMagnitude;
 
-                        var successSoFar = isMagnitudeIntegral ?
+                        bool successSoFar;
```
Set this to `true` by default. Then each operation as I'd indicated below with the suggestions needs to `&=` so an early failure doesn't get overwritten by a later success.

martincostello commented on src/JustEat.StatsD/Buffered/StatsDUtf8Formatter.cs:
```diff
@@ -79,7 +99,15 @@ private static bool TryWritePayloadWithMessageKindSuffix(ref Buffer buffer, in S
+                        bool successSoFar;
+
+                        if (msg.Operation == Operation.Increment)
+                            successSoFar = buffer.TryWriteByte((byte)'+');
```
```suggestion
                        if (msg.Operation == Operation.Increment)
                        {
                            successSoFar &= buffer.TryWriteByte((byte)'+');
                        }
```

martincostello commented on src/JustEat.StatsD/Buffered/StatsDUtf8Formatter.cs:
```diff
@@ -79,7 +99,15 @@ private static bool TryWritePayloadWithMessageKindSuffix(ref Buffer buffer, in S
+                            successSoFar = buffer.TryWriteByte((byte)'+');
+
+                        if (msg.Operation == Operation.Decrement)
+                            successSoFar = buffer.TryWriteByte((byte)'-');
```
```suggestion
                        if (msg.Operation == Operation.Decrement)
                        {
                            successSoFar &= buffer.TryWriteByte((byte)'-');
                        }
```

martincostello commented on src/JustEat.StatsD/Buffered/StatsDUtf8Formatter.cs:
```diff
@@ -79,7 +99,15 @@ private static bool TryWritePayloadWithMessageKindSuffix(ref Buffer buffer, in S
+                        if (msg.Operation == Operation.Decrement)
+                            successSoFar = buffer.TryWriteByte((byte)'-');
+
+                        successSoFar = isMagnitudeIntegral ?
```
```suggestion
                        successSoFar &= isMagnitudeIntegral ?
```

martincostello commented on src/JustEat.StatsD/IStatsDPublisher.cs:
```diff
@@ -11,21 +13,25 @@ public interface IStatsDPublisher
         /// <param name="sampleRate">The sample rate for the counter.</param>
         /// <param name="bucket">The bucket to increment the counter for.</param>
-        void Increment(long value, double sampleRate, string bucket);
+        /// <param name="tags">The key value pair collection of tags.</param>
```
```suggestion
        /// <param name="tags">An optional dictionary of tags.</param>
```

martincostello commented on src/JustEat.StatsD/IStatsDPublisher.cs:
```diff
@@ -11,21 +13,25 @@ public interface IStatsDPublisher
         /// <param name="bucket">The bucket to publish the gauge for.</param>
-        void Gauge(double value, string bucket);
+        /// <param name="operation">The gauge operation.</param>
+        /// <param name="tags">The key value pair collection of tags.</param>
```
```suggestion
        /// <param name="tags">An optional dictionary of tags.</param>
```

martincostello commented on src/JustEat.StatsD/IStatsDPublisher.cs:
```diff
@@ -11,21 +13,25 @@ public interface IStatsDPublisher
         /// <param name="sampleRate">The sample rate for the timer.</param>
         /// <param name="bucket">The bucket to publish the timer for.</param>
-        void Timing(long duration, double sampleRate, string bucket);
+        /// <param name="tags">The key value pair collection of tags.</param>
```
```suggestion
        /// <param name="tags">An optional dictionary of tags.</param>
```

martincostello commented on src/JustEat.StatsD/IStatsDPublisherExtensions.cs:
```diff
@@ -17,9 +17,10 @@ public static class IStatsDPublisherExtensions
         /// <param name="publisher">The <see cref="IStatsDPublisher"/> to publish with.</param>
         /// <param name="bucket">The bucket to increment the counter for.</param>
-        public static void Increment(this IStatsDPublisher publisher, string bucket)
+        /// <param name="tags">The key value pair collection of tags.</param>
```
```suggestion
        /// <param name="tags">An optional dictionary of tags.</param>
```

martincostello commented on src/JustEat.StatsD/IStatsDPublisherExtensions.cs:
```diff
@@ -28,9 +29,10 @@ public static void Increment(this IStatsDPublisher publisher, string bucket)
         /// <param name="value">The value to increment the counter by.</param>
         /// <param name="bucket">The bucket to increment the counter for.</param>
-        public static void Increment(this IStatsDPublisher publisher, long value, string bucket)
+        /// <param name="tags">The key value pair collection of tags.</param>
```
```suggestion
        /// <param name="tags">An optional dictionary of tags.</param>
```

martincostello commented on src/JustEat.StatsD/IStatsDPublisherExtensions.cs:
```diff
@@ -40,7 +42,8 @@ public static void Increment(this IStatsDPublisher publisher, long value, string
         /// <param name="sampleRate">The sample rate for the counter(s).</param>
         /// <param name="buckets">The bucket(s) to increment the counter(s) for.</param>
-        public static void Increment(this IStatsDPublisher publisher, long value, double sampleRate, IEnumerable<string> buckets)
+        /// <param name="tags">The key value pair collection of tags.</param>
```
```suggestion
        /// <param name="tags">An optional dictionary of tags.</param>
```

martincostello commented on src/JustEat.StatsD/IStatsDPublisherExtensions.cs:
```diff
@@ -60,7 +63,8 @@ public static void Increment(this IStatsDPublisher publisher, long value, double
         /// <param name="sampleRate">The sample rate for the counter(s).</param>
         /// <param name="buckets">The bucket(s) to increment the counter(s) for.</param>
-        public static void Increment(this IStatsDPublisher publisher, long value, double sampleRate, params string[] buckets)
+        /// <param name="tags">The key value pair collection of tags.</param>
```
```suggestion
        /// <param name="tags">An optional dictionary of tags.</param>
```

martincostello commented on src/JustEat.StatsD/IStatsDPublisherExtensions.cs:
```diff
@@ -78,9 +82,10 @@ public static void Increment(this IStatsDPublisher publisher, long value, double
         /// <param name="publisher">The <see cref="IStatsDPublisher"/> to publish with.</param>
         /// <param name="bucket">The bucket to decrement the counter for.</param>
-        public static void Decrement(this IStatsDPublisher publisher, string bucket)
+        /// <param name="tags">The key value pair collection of tags.</param>
```
```suggestion
        /// <param name="tags">An optional dictionary of tags.</param>
```

martincostello commented on src/JustEat.StatsD/IStatsDPublisherExtensions.cs:
```diff
@@ -89,9 +94,10 @@ public static void Decrement(this IStatsDPublisher publisher, string bucket)
         /// <param name="value">The value to decrement the counter by.</param>
         /// <param name="bucket">The bucket to decrement the counter for.</param>
-        public static void Decrement(this IStatsDPublisher publisher, long value, string bucket)
+        /// <param name="tags">The key value pair collection of tags.</param>
```
```suggestion
        /// <param name="tags">An optional dictionary of tags.</param>
```

martincostello commented on src/JustEat.StatsD/IStatsDPublisherExtensions.cs:
```diff
@@ -101,9 +107,10 @@ public static void Decrement(this IStatsDPublisher publisher, long value, string
         /// <param name="sampleRate">The sample rate for the counter.</param>
         /// <param name="bucket">The bucket to decrement the counter for.</param>
-        public static void Decrement(this IStatsDPublisher publisher, long value, double sampleRate, string bucket)
+        /// <param name="tags">The key value pair collection of tags.</param>
```
```suggestion
        /// <param name="tags">An optional dictionary of tags.</param>
```

martincostello commented on src/JustEat.StatsD/IStatsDPublisherExtensions.cs:
```diff
@@ -113,7 +120,8 @@ public static void Decrement(this IStatsDPublisher publisher, long value, double
         /// <param name="sampleRate">The sample rate for the counter(s).</param>
         /// <param name="buckets">The bucket(s) to decrement the counter(s) for.</param>
-        public static void Decrement(this IStatsDPublisher publisher, long value, double sampleRate, IEnumerable<string> buckets)
+        /// <param name="tags">The key value pair collection of tags.</param>
```
```suggestion
        /// <param name="tags">An optional dictionary of tags.</param>
```

martincostello commented on src/JustEat.StatsD/IStatsDPublisherExtensions.cs:
```diff
@@ -135,7 +143,8 @@ public static void Decrement(this IStatsDPublisher publisher, long value, double
         /// <param name="sampleRate">The sample rate for the counter(s).</param>
         /// <param name="buckets">The bucket(s) to decrement the counter(s) for.</param>
-        public static void Decrement(this IStatsDPublisher publisher, long value, double sampleRate, params string[] buckets)
+        /// <param name="tags">The key value pair collection of tags.</param>
```
```suggestion
        /// <param name="tags">An optional dictionary of tags.</param>
```

martincostello commented on src/JustEat.StatsD/IStatsDPublisherExtensions.cs:
```diff
@@ -156,9 +165,10 @@ public static void Decrement(this IStatsDPublisher publisher, long value, double
         /// <param name="duration">The value to publish for the timer.</param>
         /// <param name="bucket">The bucket to publish the timer for.</param>
-        public static void Timing(this IStatsDPublisher publisher, TimeSpan duration, string bucket)
+        /// <param name="tags">The key value pair collection of tags.</param>
```
```suggestion
        /// <param name="tags">An optional dictionary of tags.</param>
```

martincostello commented on src/JustEat.StatsD/IStatsDPublisherExtensions.cs:
```diff
@@ -168,9 +178,10 @@ public static void Timing(this IStatsDPublisher publisher, TimeSpan duration, st
         /// <param name="sampleRate">The sample rate for the timer.</param>
         /// <param name="bucket">The bucket to publish the timer for.</param>
-        public static void Timing(this IStatsDPublisher publisher, TimeSpan duration, double sampleRate, string bucket)
+        /// <param name="tags">The key value pair collection of tags.</param>
```
```suggestion
        /// <param name="tags">An optional dictionary of tags.</param>
```

martincostello commented on src/JustEat.StatsD/IStatsDPublisherExtensions.cs:
```diff
@@ -179,9 +190,10 @@ public static void Timing(this IStatsDPublisher publisher, TimeSpan duration, do
         /// <param name="duration">The value to publish for the timer.</param>
         /// <param name="bucket">The bucket to publish the timer for.</param>
-        public static void Timing(this IStatsDPublisher publisher, long duration, string bucket)
+        /// <param name="tags">The key value pair collection of tags.</param>
```
```suggestion
        /// <param name="tags">An optional dictionary of tags.</param>
```

martincostello commented on tests/JustEat.StatsD.Tests/statsdconfig.js:
```diff
@@ -1,4 +1,5 @@
 {
+  keyNameSanitize: false,
```
If this is needed to make tags work in general, we'd probably want this documented somewhere in the README.

Which reminds me, the README would also need to be updated to mention the tagging capabilities.

martincostello replied:
Some guidance in the need to use `keyNameSanitize` if using tags would still be useful in the README.

gandarez replied:
I think it should be always false since the tests cover tags and the StatsD server is mocked. In real life this flag does not exist, for instance when running on Graphite.

martincostello replied:
It's not a mock, it's the real thing:

https://github.com/justeat/JustEat.StatsD/blob/34f5bf6bba124cd565c52b28e8679fbf941e1e6d/.travis.yml#L27-L29

gandarez commented:
Any news here guys?

jdiazmartin commented:
I'm looking forward to having a fine-grained metrics with tags. Is there an approximate date for version 5.0.0?

martincostello commented:
> I'm looking forward to having a fine-grained metrics with tags. Is there an approximate date for version 5.0.0?

Sorry, there's no planned release date at this time. This is something we'll pick up in more detail when we have the capacity for it.

gandarez commented:
The PR will be clossed automatically soon.. no plans yet??

martincostello commented:
Hi @gandarez - sorry for the lack of movement on this PR.

For tag support, we're going to take the change from #269 as it is implemented in a more extensible way for a range of different metrics providers.

Once that is merged, if you update this PR to just suggest changes to gauges, we can re-review this PR.

Thanks!

martincostello commented:
#312 to add tagging support has been merged.
